processing/src/main/java/io/druid/query/topn/types/StringTopNColumnSelectorStrategy.java

```diff
@@ -58,15 +64,29 @@ public int getCardinality(DimensionSelector selector)
 
   @Override
   public void dimExtractionScanAndAggregate(
-      final TopNQuery query,
+      TopNQuery query,
       DimensionSelector selector,
       Cursor cursor,
       Aggregator[][] rowSelector,
       Map<Comparable, Aggregator[]> aggregatesStore
   )
   {
-    final IndexedInts dimValues = selector.getRow();
+    if (selector.getValueCardinality() != DimensionSelector.CARDINALITY_UNKNOWN) {
+      dimExtractionScanAndAggregateWithCardinality(selector, rowSelector, aggregatesStore, cursor, query);
+    } else {
+      dimExtractionScanAndAggregateNoCardinality(selector, aggregatesStore, cursor, query);
+    }
+  }
 
+  private void dimExtractionScanAndAggregateWithCardinality(
+      DimensionSelector selector,
+      Aggregator[][] rowSelector,
+      Map<Comparable, Aggregator[]> aggregatesStore,
+      Cursor cursor,
+      TopNQuery query
```

Comment:
Suggested to make `query` the first param, because 1) it is the first in `dimExtractionScanAndAggregate()` 2) it is an "input" param